usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -1083,16 +1096,19 @@ CK_RV ep11tok_pkey_check(STDLL_TokData_t *tokdata, SESSION *session,
         if (object_is_extractable(key_obj) ||
             !object_is_pkey_extractable(key_obj) ||
             object_is_attr_bound(key_obj) ||
-            !ep11tok_pkey_session_read_write(session) ||
             !ep11_data->pkey_wrap_supported) {
             goto done;
         }
+
         if (template_attribute_get_non_empty(key_obj->template,
                                              CKA_IBM_OPAQUE_PKEY,
                                              &opaque_attr) != CKR_OK ||
             !ep11tok_pkey_is_valid(tokdata, key_obj)) {
             /* this key has either no pkey attr, or it is not valid,
-             * try to create one */
+             * try to create one, if the session state allows it */
+            if (!ep11tok_pkey_session_ok_for_obj(session, key_obj))
+                goto done;
```

Comment:
Why did you move this check down? You could do the check above in line 1086 as before and would not call template_attribute_get_non_empty if the session is not ok for the obj.

Reply:
_Using an existing and valid protected key attached to a key object is not dependent on the session state._

Thats why I moved it down. You only need to check the session state if the protected key needs to be re-derived, not when it is already available and valid.